File: lib/ashy_walnut_desk/interaction/locks.ex
defmodule AshyWalnutDesk.Interaction.Locks do
  @moduledoc """
  Tiny named API for the two raw-SQL `FOR UPDATE` escape hatches the
  Interaction axis needs to defend its concurrency invariants. Today
  AshPostgres doesn't expose pessimistic locking through the Ash
  read pipeline (TO-7), so the two callsites have to hand-roll
  `Repo.query/2` + `Ecto.UUID.dump!/1` themselves.

  Consolidating behind this module:
  - Pins the raw-SQL surface to two named functions instead of two
    scattered query strings, so a future Ash-native replacement (or
    AshPostgres adding `lock_for_update`) is a single-file change.
  - Stops accidental divergence on result-shape pattern matching.
  - Documents the locking semantics in one place.

  See A3 in the simplicity review. TO-7 stays open until AshPostgres
  ships first-class pessimistic locking; this is just a tidier
  workaround.
  """

  alias AshyWalnutDesk.Repo

  @doc """
  Acquire a row-level FOR UPDATE lock on the `drafts` row identified
  by `draft_id`, filtered to `status = 'drafting'`. Returns:

  - `{:ok, :locked}` when the lock was acquired (status matched).
  - `{:ok, :not_drafting}` when the row exists but is in a different
    status (e.g. already approved by a concurrent caller).
  - `{:error, reason}` on any database error.

  Must be called inside an Ecto transaction or Ash action with
  transactional semantics; without one, the lock is released
  immediately.
  """
  @spec lock_drafting_draft(Ecto.UUID.t()) ::
          {:ok, :locked} | {:ok, :not_drafting} | {:error, term()}
  def lock_drafting_draft(draft_id) do
    draft_id_bin = Ecto.UUID.dump!(draft_id)

    case Repo.query(
           "SELECT id FROM drafts WHERE id = $1 AND status = 'drafting' FOR UPDATE",
           [draft_id_bin]
         ) do
      {:ok, %{num_rows: 1}} -> {:ok, :locked}
      {:ok, %{num_rows: 0}} -> {:ok, :not_drafting}
      {:error, _} = error -> error
    end
  end

  @doc """
  Resolve the channel_id for a draft by walking
  `drafts → inboxes → conversations`. Used by
  `CompensationAtApproval` to know which channel the approved draft
  is bound to (the `Conversation.channel_id` is authoritative).

  Returns `{:ok, channel_id_uuid}`, `{:ok, :not_found}`, or
  `{:error, reason}`.
  """
  @spec resolve_channel_for_draft(Ecto.UUID.t()) ::
          {:ok, Ecto.UUID.t()} | {:ok, :not_found} | {:error, term()}
  def resolve_channel_for_draft(draft_id) do
    draft_id_bin = Ecto.UUID.dump!(draft_id)

    case Repo.query(
           """
           SELECT c.channel_id
           FROM drafts d
           JOIN inboxes i ON i.id = d.inbox_id
           JOIN conversations c ON c.id = i.conversation_id
           WHERE d.id = $1
           """,
           [draft_id_bin]
         ) do
      {:ok, %{rows: [[channel_id]]}} -> {:ok, channel_id}
      {:ok, %{rows: []}} -> {:ok, :not_found}
      {:error, _} = error -> error
    end
  end

  @doc """
  Acquire a FOR UPDATE lock on the most-recent audit_events row for
  the given `chain_topic`, returning the hash of that row (or `nil`
  if no rows exist yet). Used by `ChainLink` to serialize hash-chain
  writes against concurrent approvers operating on the same inbox.
  """
  @spec lock_prev_audit_hash(String.t()) ::
          {:ok, String.t() | nil} | {:error, term()}
  def lock_prev_audit_hash(chain_topic) when is_binary(chain_topic) do
    sql = """
    SELECT hash
    FROM audit_events
    WHERE chain_topic = $1
    ORDER BY inserted_at DESC, id DESC
    LIMIT 1
    FOR UPDATE
    """

    case Repo.query(sql, [chain_topic]) do
      {:ok, %{rows: []}} -> {:ok, nil}
      {:ok, %{rows: [[prev_hash]]}} -> {:ok, prev_hash}
      {:error, _} = error -> error
    end
  end
end

File: crates/awd-db/src/locks.rs
//! The pessimistic `FOR UPDATE` locks — port of `interaction/locks.ex`.
//!
//! AshPostgres didn't expose pessimistic locking through its read pipeline, so
//! these were hand-rolled raw SQL in Elixir; they translate 1:1 to SQLx
//! `query`/`query_as` against a `&mut Transaction`. The SQL is pinned here as
//! constants so the (later) execution layer and tests share one source of
//! truth. Each must run inside a transaction or the lock releases immediately.

/// Lock the `drafts` row by id, only if `status = 'drafting'`. 1 row → locked;
/// 0 rows → not_drafting (or absent); used by the approve flow.
pub const LOCK_DRAFTING_DRAFT: &str =
    "SELECT id FROM drafts WHERE id = $1 AND status = 'drafting' FOR UPDATE";

/// Resolve the authoritative channel for a draft via
/// `drafts → inboxes → conversations` (the `Conversation.channel_id`).
pub const RESOLVE_CHANNEL_FOR_DRAFT: &str = "\
SELECT c.channel_id
FROM drafts d
JOIN inboxes i ON i.id = d.inbox_id
JOIN conversations c ON c.id = i.conversation_id
WHERE d.id = $1";

/// Lock the most-recent `audit_events` row for a `chain_topic` and return its
/// hash (or none). Serializes hash-chain appends against concurrent approvers
/// on the same inbox. Ordering MUST match the verifier: inserted_at DESC, id DESC.
pub const LOCK_PREV_AUDIT_HASH: &str = "\
SELECT hash
FROM audit_events
WHERE chain_topic = $1
ORDER BY inserted_at DESC, id DESC
LIMIT 1
FOR UPDATE";

/// Result of [`LOCK_DRAFTING_DRAFT`].
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DraftLock {
    Locked,
    NotDrafting,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn locks_use_for_update() {
        assert!(LOCK_DRAFTING_DRAFT.contains("FOR UPDATE"));
        assert!(LOCK_PREV_AUDIT_HASH.contains("FOR UPDATE"));
    }

    #[test]
    fn prev_hash_ordering_matches_verifier() {
        // The chain walker orders (inserted_at ASC, id ASC); the lock takes the
        // tip via DESC,DESC + LIMIT 1. These must agree on the tie-break column.
        assert!(LOCK_PREV_AUDIT_HASH.contains("ORDER BY inserted_at DESC, id DESC"));
        assert!(LOCK_PREV_AUDIT_HASH.contains("LIMIT 1"));
    }

    #[test]
    fn channel_resolution_walks_three_tables() {
        assert!(RESOLVE_CHANNEL_FOR_DRAFT.contains("JOIN inboxes"));
        assert!(RESOLVE_CHANNEL_FOR_DRAFT.contains("JOIN conversations"));
    }
}
